app/kuma-dp/pkg/dataplane/metrics/server.go

```diff
@@ -162,19 +180,54 @@ func (s *Hijacker) ServeHTTP(writer http.ResponseWriter, req *http.Request) {
 		close(out)
 	}()
 
+	// Tracks unique content type priorities returned by applications in sorted order.
+	// As there are only a few content types supported by prometheus scraper,
+	// we can use a simple array instead of a map.
+	ctPriorities := make([]bool, len(prometheusContentTypePriorities))
 	for _, app := range s.applicationsToScrape {
 		go func(app ApplicationToScrape) {
 			defer wg.Done()
-			out <- s.getStats(ctx, req, app)
+			content, contentType := s.getStats(ctx, req, app)
+			out <- content
+			if content != nil && contentType != "" {
+				// skip unknown content type
+				priorityIdx, valid := prometheusContentTypePriorities[contentType]
+				if !valid {
+					return
+				}
+				// if contentType is not found in the priorities map, it will return 0
+				ctPriorities[priorityIdx] = true
+			}
 		}(app)
 	}
 
 	select {
 	case <-ctx.Done():
 		return
 	case <-done:
-		// default format returned by prometheus
-		writer.Header().Set("content-type", string(expfmt.FmtText))
+		// We should not simply use the highest priority content type even if `application/openmetrics-text`
+		// is the superset of `text/plain`, as it might not be
+		// supported by the applications or the user might be using older prom scraper.
+		// So it's better to choose the highest negotiated content type between the apps and the scraper.
+		ct := ""
+		for priority := range ctPriorities {
+			if ctPriorities[priority] {
+				ct = prometheusPriorityContentType[priority]
+				break
+			}
+		}
+		if ct == "" {
+			// If no content type returned by applications,
+			// try to use max supported accept header, else use default.
+			acceptHeaders := strings.Split(req.Header.Get("accept"), ",")
+			ct = acceptHeaders[0]
+
+			// if accept header is not found in the priorities map, use default
+			if _, ok := prometheusContentTypePriorities[ct]; !ok {
+				ct = string(expfmt.FmtOpenMetrics_1_0_0)
```

Comment:
Is `expfmt.FmtText` not the right fallback? Why is this the default?

Reply:
Instead of falling back to `FmtText`, we are actually falling back to `FmtOpenMetrics_1_0_0`. Assuming this was a typo, after failing to get content-type from the target applications, we enter this condition where the hijacker tries to honour highest priority `Accept` type of the scraper. If the Accept type is not valid, may be because it's set to `*/*`, then we can safely fallback to `FmtOpenMetrics_1_0_0`.